src/Grpc.Net.Client/Internal/GrpcCall.cs

```diff
@@ -622,8 +640,20 @@ internal bool ResolveException(string summary, Exception ex, [NotNull] out Statu
             }
             else
             {
-                status = GrpcProtocolHelpers.CreateStatusFromException(summary, ex);
-                resolvedException = CreateRpcException(status.Value);
+                var s = GrpcProtocolHelpers.CreateStatusFromException(summary, ex);
+
+                // The server could exceed the deadline and return a CANCELLED status before the
+                // client's deadline timer is triggered. When CANCELLED is received check the
+                // deadline against the clock and change status to DEADLINE_EXCEEDED if required.
+                if (s.StatusCode == StatusCode.Cancelled &&
+                    _deadline != DateTime.MaxValue &&
+                    _deadline - Channel.Clock.UtcNow <= TimeSpan.Zero)
+                {
+                    s = new Status(StatusCode.DeadlineExceeded, s.Detail, s.DebugException);
+                }
+
+                status = s;
+                resolvedException = CreateRpcException(s);
```

Comment:
```suggestion
                resolvedException = CreateRpcException(status);
```

Nit: reads a little clearer although there's no functional difference.

Reply:
`s` is used here because `status` is nullable.